File: App/Resources/Algorithms/shellsort.js
// Same held-value-vs-live-index issue as insertionsort.js — this is a swap-based gapped insertion
// sort (using ArrayV's own ExtendedCiuraGaps sequence) rather than ArrayV's shift-based one.
function sort(engine) {
    const n = engine.count();
    const gaps = [8861, 3938, 1750, 701, 301, 132, 57, 23, 10, 4, 1];

    for (const gap of gaps) {
        if (gap >= n) continue;
        for (let i = gap; i < n; i++) {
            let j = i;
            // Strict "values[j-gap] > values[j]", same De Morgan trick as plain insertion sort.
            while (j >= gap && !engine.compare(j, j - gap)) {
                engine.swap(j, j - gap);
                j -= gap;
            }
        }
    }
}

File: App/Resources/AlgorithmDetails/shellsort.bundle/shellsort.swift
func sort(_ arr: inout [Int]) {
  let n = arr.count
  var i = n / 2
  while i > 0 {
    for j in i..<n {
      var k = j - i
      while k >= 0 {
        if arr[k + i] >= arr[k] {
          break
        } else {
          arr.swapAt(k, k + i)
        }
        k -= i
      }
    }
    i /= 2
  }
}

var array: [Int] = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(&array)
print(array)